File: leetcode.com/implement-queue-using-stacks.py
#!/bin/python3
# https://leetcode.com/problems/implement-queue-using-stacks/

class MyQueue:
    def __init__(self):
        self.stack = []
        self.is_reversed = False

    def push(self, x: int) -> None:
        if self.is_reversed:
            self.reverse()
        self.stack.append(x)

    def pop(self) -> int:
        if not self.is_reversed:
            self.reverse()

        return self.stack.pop() 

    def peek(self) -> int:
        index = -1 if self.is_reversed else 0
        return self.stack[index]

    def empty(self) -> bool:
        return len(self.stack) == 0

    def reverse(self) -> None:
        reverse_stack = []

        while self.stack:
            reverse_stack.append(self.stack.pop())

        self.is_reversed = not self.is_reversed
        self.stack = reverse_stack

File: leetcode.com/implement-queue-using-stacks.rb
class MyQueue
    def initialize()
        @stack = []
        @reversed = []
    end

=begin
    :type x: Integer
    :rtype: Void
=end
    def push(x)
        @stack << x
    end

=begin
    :rtype: Integer
=end
    def pop()
        fill_reverse()
        ret = @reverse.pop()
        reverse_back()
        return ret
    end

=begin
    :rtype: Integer
=end
    def peek()
        fill_reverse()
        ret = @reverse.last
        return ret
    end


=begin
    :rtype: Boolean
=end
    def empty()
        @stack.empty?
    end

    def fill_reverse()
        @reverse = []
        @stack.reverse_each { |x| @reverse << x}
    end

    def reverse_back()
        @stack = []
        @reverse.reverse_each { |x| @stack << x }
    end
end

# Your MyQueue object will be instantiated and called as such:
obj = MyQueue.new()
obj.push(1)
puts obj.pop()
puts obj.empty()